### PR TITLE
ROS#144: Time source implementation

### PR DESCRIPTION
**Pull request by SteelPh0enix**

### CODE REVIEW

File: src/aerugo.rs [hunk 1]
```diff
@@ -699,12 +711,31 @@ impl RuntimeApi for Aerugo {
         EVENT_MANAGER.clear()
     }
 
-    fn get_system_time(&'static self) -> crate::time::TimerInstantU64<1_000_000> {
-        Hal::get_system_time()
+    fn get_system_time(&'static self) -> Instant {
+        if let Some(system_time) = self.time_source.time_since_user_offset() {
+            return system_time;
+        }
+        TimeSource::time_since_init()
```
Comment (Glamhoth) [inline src/aerugo.rs:718]: imo `match` would be more readable here

File: src/time_source.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    /// Returns time since system initialization (call to [`Aerugo::initialize`](crate::Aerugo::initialize), start of the hardware timer)
+    pub fn time_since_init() -> Instant {
+        Hal::get_system_time()
+    }
+
+    /// Returns time since system's scheduler start (call to [`Aerugo::start`](crate::Aerugo::start)), or `None` if system hasn't started yet.
+    pub fn time_since_start(&self) -> Option<Instant> {
+        // SAFETY: This is safe as long as used in single-core context and `mark_system_start` is not being called.
+        if let Some(start_offset) = unsafe { *self.system_start_offset.as_ref() } {
```
Comment (Glamhoth) [inline src/time_source.rs:40]: imo `match` would be more readable here

File: src/time_source.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            let current_time = TimeSource::time_since_init();
+            return current_time.checked_sub_duration(start_offset);
+        }
+        None
+    }
+
+    /// Returns time since user-defined offset, or `None` if offset is not defined, or cannot be subtracted from system time.
+    pub fn time_since_user_offset(&self) -> Option<Instant> {
+        // SAFETY: This is safe as long as used in single-core context and `set_user_offset` is not being called.
+        if let Some(user_offset) = unsafe { *self.user_offset.as_ref() } {
```
Comment (Glamhoth) [inline src/time_source.rs:50]: imo `match` would be more readable here

**Review by Glamhoth — CHANGES_REQUESTED** [review]

File: src/aerugo.rs (hunk 1 above)
Comment (SteelPh0enix) [reply src/aerugo.rs]: fix'd

File: src/time_source.rs (hunk 2 above)
Comment (SteelPh0enix) [reply src/time_source.rs]: fix'd, also made better safety docs for TimeSource

File: src/time_source.rs (hunk 3 above)
Comment (SteelPh0enix) [reply src/time_source.rs]: fix'd, also made better safety docs for TimeSource

**Review by Glamhoth — APPROVED** [review]

**Pull request merged**